### PR TITLE
Yet to fix savestates

### PR DESCRIPTION
**Pull request by elad335**

### CODE REVIEW

File: rpcs3/util/serialization.hpp [hunk 1]
```diff
@@ -162,6 +162,7 @@ namespace utils
 
 				if (pos != data_offset + data.size())
 				{
+					ensure(false);
```
Comment (Megamouse) [inline rpcs3/util/serialization.hpp:165]: What's this

File: rpcs3/util/serialization.hpp [hunk 2]
```diff
@@ -108,12 +108,12 @@ namespace utils
 		{
 			if (m_is_writing) return;
 
-			const u32 offset1 = ::offset32(first) + sizeof(first);
+			const u32 offset1 = ::offset32(first) + sizeof(T);
 			const u32 offset2 = ::offset32(second);
 
-			AUDIT(offset2 >= offset1);
+			AUDIT(::offset32(first) <= ::offset32(second));
```
Comment (Megamouse) [inline rpcs3/util/serialization.hpp:114]: Why this?

File: rpcs3/util/serialization.hpp [hunk 3]
```diff
@@ -108,12 +108,12 @@ namespace utils
...
 
-			if (offset2 > offset1)
+			if (::offset32(first) <= ::offset32(second))
```
Comment (Megamouse) [inline rpcs3/util/serialization.hpp:116]: Why this? this makes no sense

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 4]
```diff
@@ -479,10 +479,6 @@ struct gem_config_data
 	gem_config_data(utils::serial& ar)
 	{
 		save(ar);
-
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp:482]: Why remove this?

File: rpcs3/util/serialization.hpp (hunk 2 above)
Comment (elad335) [reply rpcs3/util/serialization.hpp:114]: This is because of c++20's [[no_unique_address]].

File: rpcs3/util/serialization.hpp (hunk 3 above)
Comment (elad335) [reply rpcs3/util/serialization.hpp:116]: This is because of c++20's [[no_unique_address]].

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 4 above)
Comment (elad335) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp:482]: It's not needed

File: rpcs3/util/serialization.hpp (hunk 3 above)
Comment (Megamouse) [reply rpcs3/util/serialization.hpp:116]: It doesn't make sense still.
And your explanation doesn't tell me anything

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 4 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp:482]: Why is it not needed?

Comment (elad335) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp:482]: Because ar is in reading state in a constructor

File: rpcs3/util/serialization.hpp (hunk 3 above)
Comment (elad335) [reply rpcs3/util/serialization.hpp:116]: This attribute allows two distinguished members to have the same offset,

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 4 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp:482]: Why is the function called save then?

File: rpcs3/util/serialization.hpp (hunk 3 above)
Comment (Megamouse) [reply rpcs3/util/serialization.hpp:116]: It still makes no sense to change this condition if you look at the code in the scope

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 4 above)
Comment (elad335) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp:482]: It recycles the function, but it doesn't save here

File: rpcs3/util/serialization.hpp (hunk 3 above)
Comment (elad335) [reply rpcs3/util/serialization.hpp:116]: I see, I only need to change the assert.

**Comment by Darkhost1999** [comment]
[RPCS3.log](https://github.com/user-attachments/files/18245673/RPCS3.log)
Tested the last 3 pushes starting 1 hour ago this is the most recent test with a typo fix for offset instead of offse.
The error
```
·F 0:00:52.733079 {Emu State Load Thread: 'class ppu_function_manager'} SIG: Thread terminated due to fatal error: serial_breathe_and_tag(1): { reading, 0x8d207a8/0x8d432d4, memory=0x95b47d, sample64=0xc52e8bc300000000 }
object: 'none', next-object: 'class ppu_function_manager', expected/tag: 0x3d3 != 0xd4,
(in file C:\Users\bryso\Documents\GitHub\rpcs3\rpcs3\Emu\savestate_utils.cpp:396[:24], in function 'ushort stx::serial_breathe_and_tag()')
·F 0:00:52.733106 {Emu State Load Thread: 'class ppu_function_manager'} profiler: cpu_thread::flush_profilers() has been called incorrectly.
```

**Comment by elad335** [comment]
@Darkhost1999 it's not meant for already broken saves, only for newer saves.

**Comment by Darkhost1999** [comment]
I clicked save state to create a state. But I'll clean the output folder just in case

File: rpcs3/util/serialization.hpp [hunk 5]
```diff
@@ -459,10 +459,16 @@ namespace utils
...
 		{
-			ensure(data.size() + data_offset >= backwards);
-			pos = data.size() + data_offset - backwards;
+			pos = data.size() + data_offset;
+			return pos;
+		}
+
+		usz trunc(usz count)
+		{
+			data.resize(data.size() - count);
```
Comment (Megamouse) [inline rpcs3/util/serialization.hpp:470]: Why no clamp here?

**Comment by Darkhost1999** [comment]
The error has changed
```
E SYS: SPU thread continued after being paused. (old_pc=0x11a8, pc=0xcf4, unsavable=1)
F {Emu State Capture Thread: ''} SIG: Thread terminated due to fatal error: Verification failed (object: 0x0)
(in file D:\a\1\s\rpcs3\Emu\Cell\Modules\cellSysutil.cpp:119[:3], in function 'void sysutil_cb_manager::save(utils::serial &)')
E SYS: Saving savestate failed due to fatal error!
```
[RPCS3.log](https://github.com/user-attachments/files/18246668/RPCS3.log)
I tried Dark Void, Lego Harry Potter, and Bioshock

**Comment by Apopoy** [comment]
This update still not fix in wwe 2k14 referee still have bugs issue even I already turned on Write buffers

[RPCS3.log.gz](https://github.com/user-attachments/files/18265061/RPCS3.log.gz)

**Pull request merged**